Vault standard; SRC-6 (#24)
FuelLabs/sway-standards: standards/src_6/README.md

<p align="center">
    <picture>
        <source media="(prefers-color-scheme: dark)" srcset=".docs/src-5-logo-dark-theme.png">
        <img alt="SRC-5 logo" width="400px" src=".docs/src-5-logo-light-theme.png">
    </picture>
</p>

> # Abstract

Reviewer: ```suggestion
# Abstract

```
Same for all the other headers

The following standard allows for the implementation of a standard API for token vaults such as yield bearing token vaults. This standard is an optional add-on to the SRC-20 standard.

# Motivation

> Token vaults allow users to own shares of variable amount of assets, such as lending protocols which may have growing assets due to profits from interest. This pattern is highly useful and would greatly benefit from standardisation

Reviewer 2: ```suggestion

Token vaults allow users to own shares of variable amount of assets, such as lending protocols which may have growing assets due to profits from interest. This pattern is highly useful and would greatly benefit from standardisation.
```

# Prior Art
Token vaults have been thoroughly explored on Ethereum and with [EIP 4626](https://eips.ethereum.org/EIPS/eip-4626) they have their own standard for it. However as Fuel's native assets are fundamentally different to Ethereum's ERC-20 tokens, the implementation will differ, but the interface may be used as reference.

# Specification
## Required public functions
The following functions MUST be implemented (on top of the SRC-20 functions) to follow the SRC-6 standard

### `fn deposit(receiver: Identity) -> u64`
Method that allows depositing of the underlying asset in exchange for shares of the vault.

> This function takes the receiver's identity as an argument and returns the amount of shares minted.

Reviewer 2: Instead of `returns` I would explicitly mention that it transfers the shares of some sub id or whatever. `returns` implies it may be just a number returned from a contract call (getter)

This function takes the asset's AssetId and the receiver's identity as arguments and returns the amount of assets transferred.
The AssetId of the asset, and the AssetId of the shares MUST be one-to-one, meaning every deposited AssetId shall have a unique corresponding shares AssetId.

MUST revert if any AssetId other than the AssetId corresponding to the deposited asset is forwarded.
MUST send `preview_withdraw(redeemed_shares)` amount of assets to `receiver`.
MUST burn the received shares.
MUST reduce `managed_assets` by `preview_withdraw(redeemed_shares)`.
MUST reduce `total_supply` of the shares's AssetId by amount of burnt shares.
MUST emit a `Withdraw` log.

### `fn managed_assets(asset: AssetId) -> u64`
Method that returns the total assets under management by vault. Includes assets controlled by the vault but not directly possessed by vault.
This function takes the asset's AssetId as an argument and returns the total amount of assets of AssetId under management by vault.

MUST return total amount of assets of underlying AssetId under management by vault.
MUST return 0 if there are no assets of underlying AssetId under management by vault.

> MUST NOT revert under any circumstances.

Reviewer 2: I believe there is documentation below which states that it may revert...

Author: Where?

Reviewer 2: Without having to look over all the work I imagine I've look at the docs in the code itself

Author: Okay well i didnt find any such documentation so im just gonna assume you misread something unless you manage to find it

### `fn convert_to_shares(asset: AssetId, assets: u64) -> Option<u64>`
Helper method for converting assets to shares.
This function takes the asset's AssetId and the amount of assets as arguments and returns the amount of shares that would be minted for the given amount of assets, in an ideal condition without slippage.

MUST return an Option::Some of the amount of shares that would be minted for the given amount of assets, without accounting for any slippage, if the given asset is supported.
MUST return an Option::None if the given asset is not supported.
MUST NOT revert under any circumstances.

### `fn convert_to_assets(asset: AssetId, shares: u64) -> Option<u64>`
Helper method for converting shares to assets.
This function takes the asset's AssetId and the amount of shares as arguments and returns the amount of assets that would be transferred for the given amount of shares, in an ideal condition without slippage.

MUST return an Option::Some of the amount of assets that would be transferred for the given amount of shares, if the given asset is supported.
MUST return an Option::None if the asset is not supported.
MUST NOT revert under any circumstances.

> ### `fn preview_deposit(asset: AssetId, assets: u64) -> u64`

Reviewer 3: Will this always be the same as `convert_to_shares`?

Author: No, convert_to_shares wont account for any slippage and will not revert. Basically ideal conditions. the preview functions can revert for any reason the main function will revert, and they return the amount of assets/shares that would be received if it was a withdraw/deposit call instead (in the exact same context and conditions)

Helper method for previewing deposit.
This function takes the asset's AssetId and the amount of assets as arguments and returns the amount of shares that would have been minted for the given amount of assets.

MUST return the amount of shares that would have been minted for the given amount of assets.
MUST revert for any reason the `deposit` function would revert given the same conditions.

> ### `fn preview_withdraw(asset: AssetId, shares: u64) -> u64`

Reviewer 3: Will this always be the same as `convert_to_assets`?

This function takes the asset's AssetId and the amount of shares as arguments and returns the amount of assets that would have been transferred for the given amount of shares.

MUST return the amount of assets that would have been transferred for the given amount of shares.
MUST revert for any reason the `withdraw` function would revert given the same conditions.

### `fn max_depositable(asset: AssetId) -> Option<u64>`
Helper method for getting maximum depositable
This function takes the asset's AssetId as an argument and returns the maximum amount of assets that can be deposited into the contract, for the given asset.

MUST return the maximum amount of assets that can be deposited into the contract, for the given asset.

### `fn max_withdrawable(asset: AssetId) -> Option<u64>`
Helper method for getting maximum withdrawable
This function takes the asset's AssetId as an argument and returns the maximum amount of assets that can be withdrawn from the contract, for the given asset.

MUST return the maximum amount of assets that can be withdrawn from the contract, for the given asset.

## Required logs
The following logs MUST be emitted at the specified occasions

> ```

Reviewer 2: ```suggestion
```rust
```

/// Event logged when a deposit is made.
pub struct Deposit {
    /// The caller of the deposit function.
    caller: Identity,
    /// The receiver of the deposit.
    receiver: Identity,
    /// The asset being deposited.
    asset: AssetId,
    /// The amount of assets being deposited.
    assets: u64,
    /// The amount of shares being minted.
    shares: u64,
}
```

> `caller` has called the `deposit` method sending `assets` assets of the `asset` AssetId, in exchange for `shares` shares sent to the receiver `receiver`

Reviewer 2: ```suggestion

`caller` has called the `deposit` method sending `assets` assets of the `asset` AssetId, in exchange for `shares` shares sent to the receiver `receiver`
```

The `Deposit` struct MUST be logged whenever new shares are minted via the `deposit` method

> ```

Reviewer 2: ```suggestion
```rust
```

/// Event logged when a withdrawal is made.
pub struct Withdraw {
    /// The caller of the withdrawal function.
    caller: Identity,
    /// The receiver of the withdrawal.
    receiver: Identity,
    /// The asset being withdrawn.
    asset: AssetId,
    /// The amount of assets being withdrawn.
    assets: u64,
    /// The amount of shares being burned.
    shares: u64,
}
```

> `caller` has called the `withdraw` method sending `shares` shares in exchange for `assets` assets of the `asset` AssetId to the receiver `receiver`

Reviewer 2: ```suggestion

`caller` has called the `withdraw` method sending `shares` shares in exchange for `assets` assets of the `asset` AssetId to the receiver `receiver`
```

The `Withdraw` struct MUST be logged whenever shares are redeemed for assets via the `withdraw` method

# Rationale

> The ABI discussed is simple and covers the known use cases of token vaults while allowing safe implementations

Reviewer 2: ```suggestion

The ABI discussed covers the known use cases of token vaults while allowing safe implementations
```

# Backwards compatibility
This standard is fully compatible with the SRC-20 standard

# Security Considerations

> Incorrect implementation of token vaults could allow attackers to steal underlying assets. It is recommended to properly audit any code using this standard to ensure exploits are not possible.

Reviewer 3: I wonder if we should include something about how a vault can "lie" about the underlying?

Ex, if a vault says some asset represents USDC, there's no guarantee that the vault _actually_ represents USDC (may be a consideration for wallets)

This function takes the asset's AssetId, the sub_id of the sub vault, and the receiver's identity as arguments and returns the amount of assets transferred to the receiver.
The AssetId of the asset, and the AssetId of the shares MUST be one-to-one, meaning every deposited AssetId shall have a unique corresponding shares AssetId.

MUST revert if any AssetId other than the AssetId corresponding to the deposited asset is forwarded.
MUST burn the received shares.
MUST reduce `total_supply` of the shares's AssetId by amount of burnt shares.
MUST emit a `Withdraw` log.

### `fn managed_assets(asset: AssetId, sub_id: SubId) -> u64`
Method that returns the total assets under management by vault. Includes assets controlled by the vault but not directly possessed by vault.
This function takes the asset's AssetId and the sub_id of the sub vault as an argument and returns the total amount of assets of AssetId under management by vault.

MUST return total amount of assets of underlying AssetId under management by vault.
MUST return 0 if there are no assets of underlying AssetId under management by vault.
MUST NOT revert under any circumstances.

> ### `fn convert_to_shares(asset: AssetId, sub_id: SubId, assets: u64) -> Option<u64>`

Reviewer 3: Can this be removed in favor of just simulating deposit?

<p align="center">
    <picture>
        <source media="(prefers-color-scheme: dark)" srcset=".docs/src-5-logo-dark-theme.png">
        <img alt="SRC-5 logo" width="400px" src=".docs/src-5-logo-light-theme.png">
    </picture>
</p>

# Abstract

The following standard allows for the implementation of a standard API for token vaults such as yield bearing token vaults. This standard is an optional add-on to the SRC-20 standard.

# Motivation

Token vaults allow users to own shares of variable amount of assets, such as lending protocols which may have growing assets due to profits from interest. This pattern is highly useful and would greatly benefit from standardisation

# Prior Art

Token vaults have been thoroughly explored on Ethereum and with [EIP 4626](https://eips.ethereum.org/EIPS/eip-4626) they have their own standard for it. However as Fuel's native assets are fundamentally different to Ethereum's ERC-20 tokens, the implementation will differ, but the interface may be used as reference.

# Specification

> ## Required public functions

Reviewer: ```suggestion
## Required public functions

```
Add new lines below as well

MUST revert if any AssetId other than the underlying is forwarded.
MUST mint `preview_deposit(deposited_assets)` amount of shares to `receiver`.
MUST increase `managed_assets` by `deposited_assets` (through any means including `std::context::this_balance(ASSET_ID)` if applicable).
MUST increase `total_supply` of the share's AssetId by newly minted shares.
MUST increase `total_assets` by one if the the AssetId is minted for the first time.
MUST emit a `Deposit` log.

### `fn withdraw(asset: AssetId, sub_id: SubId, receiver: Identity) -> u64`
Method that allows the redeeming of the vault shares in exchange for a pro-rata amount of the underlying asset
This function takes the asset's AssetId, the sub_id of the sub vault, and the receiver's identity as arguments and returns the amount of assets transferred to the receiver.
The AssetId of the asset, and the AssetId of the shares MUST be one-to-one, meaning every deposited AssetId shall have a unique corresponding shares AssetId.

MUST revert if any AssetId other than the AssetId corresponding to the deposited asset is forwarded.

> MUST send `preview_withdraw(redeemed_shares)` amount of assets to `receiver`.

Reviewer: What is `preview_withdraw`????

MUST return an Option::None if the given asset is not supported.
MUST NOT revert under any circumstances.

### `fn convert_to_assets(asset: AssetId, sub_id: SubId, shares: u64) -> Option<u64>`
Helper method for converting shares to assets.
This function takes the asset's AssetId, the sub_id of the sub vault, and the amount of shares as arguments and returns the amount of assets that would be transferred for the given amount of shares, in an ideal condition without slippage.

MUST return an Option::Some of the amount of assets that would be transferred for the given amount of shares, if the given asset is supported.
MUST return an Option::None if the asset is not supported.
MUST NOT revert under any circumstances.

### `fn max_depositable(asset: AssetId, sub_id: SubId) -> Option<u64>`
Helper method for getting maximum depositable
This function takes the asset's AssetId and the sub_id of the sub vault as an argument and returns the maximum amount of assets that can be deposited into the contract, for the given asset.

MUST return the maximum amount of assets that can be deposited into the contract, for the given asset.

> ### `fn max_withdrawable(asset: AssetId, sub_id: SubId) -> Option<u64>`

Reviewer: Since this is just a helper function I doubt whether this should even be included

Author: Wdym? Its a abi function

Helper method for getting maximum withdrawable
This function takes the asset's AssetId and the sub_id of the sub vault as an argument and returns the maximum amount of assets that can be withdrawn from the contract, for the given asset.

MUST return the maximum amount of assets that can be withdrawn from the contract, for the given asset.

### `fn vault_asset_id(asset: AssetId, sub_id: SubId) -> Option<AssetId>`
Method that returns the AssetId of the vault shares for the given asset and sub vault.
This function takes the asset's AssetId and the SubId of the vault as arguments and returns the AssetId of the vault shares for the given asset and sub vault.

> MUST return an Option::Some of the AssetId of the vault shares for the given asset and sub vault, if the given asset is supported.

Reviewer: ```suggestion
MUST return `Some(AssetId)` of the vault shares for the given asset and sub vault, if the given asset is supported.
```
This should be applied to multiple lines

MUST return an Option::None if the given asset is not supported.
MUST NOT revert under any circumstances.

### `fn asset_of_vault(vault_asset_id: AssetId) -> Option<AssetId>`
Method that returns the AssetId of the asset of the vault for the given AssetId of the vault shares.
This function takes the AssetId of the vault shares as an argument and returns the AssetId of the asset of the vault for the given AssetId of the vault shares.

MUST return an Option::Some of the AssetId of the asset of the vault for the given AssetId of the vault shares, if the given asset is supported and the vault has been initialised.
MUST return an Option::None if the given asset is not supported or the vault has not been initialised.
MUST NOT revert under any circumstances.

## Required logs
The following logs MUST be emitted at the specified occasions

```sway

> /// Event logged when a deposit is made.

Author: Im confused, thats also just the specification enum, just without comments

    /// The SubId of the vault.
    sub_id: SubId,
    /// The amount of assets being deposited.
    assets: u64,
    /// The amount of shares being minted.
    shares: u64,
}
```
`caller` has called the `deposit` method sending `assets` assets of the `asset` AssetId to the subvault of `sub_id`, in exchange for `shares` shares sent to the receiver `receiver`

The `Deposit` struct MUST be logged whenever new shares are minted via the `deposit` method

```sway
/// Event logged when a withdrawal is made.
pub struct Withdraw {
    /// The caller of the withdrawal function.
    caller: Identity,
    /// The receiver of the withdrawal.
    receiver: Identity,
    /// The asset being withdrawn.
    asset: AssetId,
    /// The SubId of the vault.
    sub_id: SubId,
    /// The amount of assets being withdrawn.
    assets: u64,
    /// The amount of shares being burned.
    shares: u64,
}
```
`caller` has called the `withdraw` method sending `shares` shares in exchange for `assets` assets of the `asset` AssetId from the subvault of `sub_id` to the receiver `receiver`

The `Withdraw` struct MUST be logged whenever shares are redeemed for assets via the `withdraw` method

# Rationale

> The ABI discussed is simple and covers the known use cases of token vaults while allowing safe implementations

Reviewer: ```suggestion
The ABI discussed covers the known use cases of token vaults while allowing safe implementations.
```

> # Backwards compatibility

Reviewer: ```suggestion
# Backwards Compatibility
```

This standard is fully compatible with the SRC-20 standard

# Security Considerations

Incorrect implementation of token vaults could allow attackers to steal underlying assets. It is recommended to properly audit any code using this standard to ensure exploits are not possible.

> # Reference implementation
> 
> Full reference implementation can be seen [here](https://github.com/SwayStar123/vault-standard-reference-implementation/blob/master/src/main.sw)
> 

Reviewer: ```suggestion
# Example ABI

sway
abi SRC6 {
/// Fill in
}

# Example Implementation

## Single (link)

Desc.

## Muti (link)

Desc.
```

This function MUST emit a `Deposit` log.
This function MUST return the amount of minted shares.

### `fn withdraw(receiver: Identity, asset: AssetId, sub_id: SubId) -> u64`

Method that allows the redeeming of the vault shares in exchange for a pro-rata amount of the underlying asset
This function takes the asset's AssetId, the sub_id of the sub vault, and the receiver's identity as arguments and returns the amount of assets transferred to the receiver.
The AssetId of the asset, and the AssetId of the shares MUST be one-to-one, meaning every deposited AssetId shall have a unique corresponding shares AssetId.

This function MUST revert if any AssetId other than the AssetId representing the deposited asset's shares for the given sub vault at `sub_id` is forwarded.
This function MUST burn the received shares.
This function MUST emit a `Withdraw` log.
This function MUST return amount of assets transferred to the receiver.

### `fn managed_assets(asset: AssetId, sub_id: SubId) -> u64`

> Method that returns the total assets under management by vault. Includes assets controlled by the vault but not directly possessed by vault.

Author: Vault strategies can utilize deposited assets in other dapps and such

This function MUST return an `Option::Some(amount)` if the given vault exists.
This function MUST return an `Option::None` if the given vault does not exist.
This function MUST account for both global and user specific limits. For example: if deposits are disabled, even temporarily, MUST return 0.

### `fn max_withdrawable(receiver: Identity, asset: AssetId, sub_id: SubId) -> Option<u64>`

Helper method for getting maximum withdrawable
This function takes the hypothetical receive's `Identity`, the asset's `AssetId`, and the `sub_id`` of the sub vault as an argument and returns the maximum amount of assets that can be withdrawn from the contract, for the given asset.

This function MUST return the maximum amount of assets that can be withdrawn from the contract, for the given asset, if the given vault exists.
This function MUST return an `Option::Some(amount)` if the given vault exists.
This function MUST return an `Option::None` if the given vault does not exist.
This function MUST account for global limits. For example: if withdrawals are disabled, even temporarily, MUST return 0.

## Required logs

The following logs MUST be emitted at the specified occasions

```sway

> /// Event logged when a deposit is made.

Reviewer: ```suggestion
### Deposit

The `Deposit` log SHALL have the following fields.

#### - `caller`: Identity

The `caller` field MUST represent the Identity which called the deposit function.

#### - `receiver`: Identity

The `receiver` field MUST represent the Identity which received the vault shares. 

#### - `asset`: AssetId

The `asset` field MUST represent the AssetId of the asset which was deposited into the vault.

#### - `sub_id`: SubId

The `sub_id` field MUST represent the SubId of the vault which was deposited into.

#### - `deposit_amount`: u64

The `deposit_amount` field MUST represent the u64 amount of assets deposited into the vault.

#### - `shares`: u64

The `shares` field MUST represent the u64 amount of shares minted.
```

pub struct Deposit {
    /// The caller of the deposit function.
    caller: Identity,
    /// The receiver of the deposit.
    receiver: Identity,
    /// The asset being deposited.
    asset: AssetId,
    /// The SubId of the vault.
    sub_id: SubId,
    /// The amount of assets being deposited.
    assets: u64,
    /// The amount of shares being minted.
    shares: u64,
}
```

> `caller` has called the `deposit` method sending `assets` assets of the `asset` AssetId to the subvault of `sub_id`, in exchange for `shares` shares sent to the receiver `receiver`

Reviewer: This can be removed if you define the struct in the comment above

> The `Deposit` struct MUST be logged whenever new shares are minted via the `deposit` method

Reviewer: This can be moved under the ### Deposit header or remove since it's defined in the `deposit()` function.

```sway

> /// Event logged when a withdrawal is made.

Reviewer: ```suggestion
### Withdraw

The `Withdraw` log SHALL have the following fields.

#### - `caller`: Identity

The `caller` field MUST represent the Identity which called the withdraw function.

#### - `receiver`: Identity

The `receiver` field MUST represent the Identity which received the withdrawn assets.

#### - `withdawn_asset`: AssetId

The `deposit_asset` field MUST represent the AssetId of the asset that was withdrawn.

#### - `withdrawn_id`: SubId

The `vault_id` field MUST represent the SubId of the vault from which was withdrawn.

#### - `withdraw_amount`: u64

The `withdraw_amount` field MUST represent the u64 amount of tokens withdrawn.

#### - `shares`: u64

The `shares` field MUST represent the u64 amount of shares burned.

```

pub struct Withdraw {
    /// The caller of the withdrawal function.
    caller: Identity,
    /// The receiver of the withdrawal.
    receiver: Identity,
    /// The asset being withdrawn.
    asset: AssetId,
    /// The SubId of the vault.
    sub_id: SubId,
    /// The amount of assets being withdrawn.
    assets: u64,
    /// The amount of shares being burned.
    shares: u64,
}
```

> `caller` has called the `withdraw` method sending `shares` shares in exchange for `assets` assets of the `asset` AssetId from the subvault of `sub_id` to the receiver `receiver`

Reviewer: This can be removed

> The `Withdraw` struct MUST be logged whenever shares are redeemed for assets via the `withdraw` method

Reviewer: Removed or move up to the ### Deposit section

# Rationale

> The ABI discussed and covers the known use cases of token vaults while allowing safe implementations

Reviewer: ```suggestion
The ABI discussed covers the known use cases of token vaults while allowing safe implementations.
```

# Backwards Compatibility

> This standard is fully compatible with the SRC-20 standard

Reviewer: ```suggestion
This standard is fully compatible with the [SRC-20 standard](https://github.com/FuelLabs/sway-standards/tree/master/standards/src_20).
```

<p align="center">
    <picture>
        <source media="(prefers-color-scheme: dark)" srcset=".docs/src-6-logo-dark-theme.png">
        <img alt="SRC-6 logo" width="400px" src=".docs/src-6-logo-light-theme.png">
    </picture>
</p>

# Abstract

The following standard allows for the implementation of a standard API for token vaults such as yield bearing token vaults. This standard is an optional add-on to the SRC-20 standard.

# Motivation

Token vaults allow users to own shares of variable amount of assets, such as lending protocols which may have growing assets due to profits from interest. This pattern is highly useful and would greatly benefit from standardisation

# Prior Art

> Token vaults have been thoroughly explored on Ethereum and with [EIP 4626](https://eips.ethereum.org/EIPS/eip-4626) they have their own standard for it. However as Fuel's native assets are fundamentally different to Ethereum's ERC-20 tokens, the implementation will differ, but the interface may be used as reference.

Reviewer: ```suggestion
Token vaults have been thoroughly explored on Ethereum and with [EIP 4626](https://eips.ethereum.org/EIPS/eip-4626) they have their own standard for it. However as Fuel's [Native Assets](https://docs.fuel.network/docs/sway/blockchain-development/native_assets) are fundamentally different from Ethereum's ERC-20 tokens, the implementation will differ, but the interface may be used as a reference.
```

# Specification

## Required public functions

The following functions MUST be implemented to follow the SRC-6 standard. Any contract that implements the SRC-6 standard MUST implement the SRC-20 standard.

### `fn deposit(receiver: Identity, sub_id: SubId) -> u64`

> Method that allows depositing of the underlying asset in exchange for shares of the vault.

Reviewer: ```suggestion
This function MUST allow for depositing of the underlying asset in exchange for shares of the vault.
```

> This function takes the receiver's identity and the sub_id of the sub vault as an argument and returns the amount of shares minted to the receiver.

Reviewer: ```suggestion
This function takes the `receiver` Identity and the SubId `sub_id` of the sub-vault as an argument and returns the amount of shares minted to the receiver.
```

> MUST revert if any unaccepted AssetId is forwarded.

Reviewer: ```suggestion
This function MUST revert if any unaccepted asset for a particular sub-vault is sent.
```

> MUST increase `managed_assets` by amount of deposited assets (through any means including `std::context::this_balance(ASSET_ID)` if applicable).

Reviewer: ```suggestion
This function MUST increment `managed_assets` by the amount of deposited assets (through any means including `std::context::this_balance(ASSET_ID)` if applicable).
```
What is `managed_assets`. Reading chronologically through this standard this is the first time I'm coming across this. Are we standardizing variables?

Could we state that this function instead 
`This function must keep track of deposits` instead?

> MUST mint a token representing the pro-rata share of the vault, with the AssetId of `sha256((asset, sub_id))`, a hash of the AssetId of the deposited asset, and the `sub_id` of the vault.

Reviewer: ```suggestion
This function MUST mint an asset representing the pro-rata share of the vault, with the SubId of the `sha256((asset, sub_id))` digest, where `asset` is the AssetId of the deposited asset and the `sub_id` is the id of the vault.
```

> MUST increase `total_supply` of the share's AssetId by newly minted shares.

Reviewer: ```suggestion
This function MUST increase the [SRC-20](https://github.com/FuelLabs/sway-standards/tree/master/standards/src_20) total supply of the sub-vault's asset.
```

> MUST increase `total_assets` by one if the the AssetId is minted for the first time.

Reviewer: ```suggestion
MUST increase the [SRC-20](https://github.com/FuelLabs/sway-standards/tree/master/standards/src_20) total assets by one if the sub-vault is minted for the first time.
```

Author: doesnt make sense to say a sub vault is minted. Is "sub-vault share asset" too verbose?

> MUST emit a `Deposit` log.

Reviewer: ```suggestion
This function MUST emit a `Deposit` log.
```

> MUST return amount of minted shares.

Reviewer: ```suggestion
This function MUST return the amount of minted shares.
```
